Fix restart after install

feerrenrut opened the pull request:
Note to reviewers, it will be easiest to review this commit-wise, separate groups of changes are in individual commits. I believe this is clean enough to be merged rather than squash merged.

Please consult individual commit messages which contain descriptions of each of the changes.

### Link to issue number:
Fixes #12503

### Summary of the issue:
On my machine, I can reproduce the issue most of time, but not always. Extra logging has helped me to verify the outcome.
With mouse tracking and play audio coordinates enabled in config, I run the launcher with `--install -m` arguments, and in the final stages of the installer move the mouse a lot to slow down the exiting process.

The main issue was introduced with PR #12431 f8852a8, prior to this `doInstall` in `installerGui.py` would just start a new copy of NVDA allowing the new copy of NVDA to terminate the old copy by posting a `WM_QUIT` message.

With the update of wxPython, this caused a crash, see PR #12431 for details.

The new approach introduced in #12431 was to trigger the exit NVDA after starting the new process.
This leads to a race condition for ownership of the "single NVDA instance" mutex.

### Description of how this pull request fixes the issue:

To fix this race condition the mutex is created without taking ownership, then `waitForSingleObject` is used to take ownership.
If the old NVDA process is still running, the new process waits.
If the old NVDA process has already released the mutex, the new process will take ownership immediately.
If the old NVDA process has crashed and not released the mutex, the new process takes ownership and logs the situation.

Note, previously only `CloseHandle` was used, this relied on the process exiting and Windows releasing the mutex.
Now, the mutex is explicitly released before the handle is closed.

Useful links:
- [Handling Abandoned mutexes - Raymond Chen](https://devblogs.microsoft.com/oldnewthing/20140925-00/?p=43983)
- [Mutex objects overview](https://docs.microsoft.com/en-us/windows/win32/sync/mutex-objects)
- [CreateMutex Docs](https://docs.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-createmutexw)
- [WaitForSingleObject Docs](https://docs.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-waitforsingleobject)
- [ReleaseMutex Docs](https://docs.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-releasemutex)
- [CloseHandle Docs](https://docs.microsoft.com/en-us/windows/win32/api/handleapi/nf-handleapi-closehandle)

### Testing strategy:
Due to the intermittent nature of the problem, post fix I used extra logging (not included with PR) to verify when race condition triggered and successfully avoided.

I tested:
- Running from source
- Installing with `--install -m`
- Updating from `nvda_snapshot_alpha-23006,a5a5f2f0.exe` (having saved a `updateCheckState.pickle` file previously, and renaming my source built launcher to match the expected NVDA update file)
- Running from installed copy

This will require testing on alpha. Note to testers, please thoroughly look at this. We aim to get this into beta as soon as possible.

### Known issues with pull request:
I believe there is still a race condition between exiting process cleaning up the GUI and the new process calling `FindWindow`

### Change log entries:
None

### Code Review Checklist:
- [x] Pull Request description is up to date.
- [x] Unit tests.
- [x] System (end to end) tests.
- [x] Manual testing.
- [x] User Documentation.
- [x] Change log entry.
- [x] Context sensitive help for GUI changes.
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers

feerrenrut committed:
Fix conflation of silent and update

Fix an issue introduced in commit 93bbd21b77e1596309a67a4e4fc291b9f3af7589
"Add --copy-portable-config command line parameter to automatically copy
 the portable configuration with a silent installation #9679"

feerrenrut committed:
Clarify log message for trigger NVDA exit

There is only an error case when GUI intiated exiting is not first
to trigger the exit.

Previously there was a warning for every exit, since there is
always a "catch-all" exit trigger.
The "catch-all" exit trigger is necessary to ensure there cleanup happens
when a WM_QUIT windows message is recieved, causing wx to exit the
MainLoop.
Now there is only a message when the "catch-all" is first, aimed at helping
developers identify when exit was due to WM_QUIT.

feerrenrut committed:
Maintain logging level on restart

When debugging installer or restart behavior, maintaining the
commandline provided log level is helpful.

Other wise the logging level must be set via config,
this happens much later in the startup of NVDA resulting in many
debug level log messages being lost.

feerrenrut committed:
Prevent race condition when restarting NVDA

Prior to f8852a8482a353f415c1070530d06bc6f7b054ec doInstall in
installerGui.py would just start a new copy of NVDA allowing the
new copy of NVDA to terminate the old copy by posting a WM_QUIT message.

With the update of wxPython, this caused a crash, see PR #12431 for
details.
The new approach was to trigger the exit NVDA after starting the new
process.
This leads to a race condition for ownership of the "single NVDA
instance" mutex.

To fix this race condition the mutex is created without taking ownership,
then waitForSingleObject is used to take ownership.
If the old NVDA process is still running, the new process waits.
If the old NVDA process has already released the mutex, the new process
will take ownership immediately.
If the old NVDA process has crashed and not released the mutex, the new
process takes ownership and logs the situation.

Note, previously only CloseHandle was used, this relied on the process
exiting and Windows releasing the mutex.
Now, the mutex is explicitly released before the handle is closed.

feerrenrut committed:
Logging: detecting old NVDA process via FindWindow

This logging won't be seen unless the logHandler initialisation is moved
to earlier in this file.
This has been helpful during development, but changes to the logging
are not necessary for this change.

seanbudd commented:
Converted to draft due to system test failure.

michaelDCurran commented on source/nvda.pyw:
```diff
@@ -213,12 +230,83 @@ def getDesktopName():
+				" Note: Restarting your system is recommended."
+				" This error indicates that NVDA previously did not exit correctly or was terminated"
+				" (perhaps by the task manager)."
+				" Your system is in an indeterminate state."
```
This message really botheres me. I think the reboot part is enough. The system being in an indeterminant state may scare some beginner users more than necesary.

feerrenrut replied:
It's true though right, NVDA crashing could leave the system in an unknown state. The word choice was intended to scare users into restarting. Perhaps it would be more helpful (at least for developers) if the key unknowns were listed?

feerrenrut replied:
I've removed this line from the message.

feerrenrut committed:
Try to capture py2exe log

This log is created by py2exe when somemthing writes to stdErr.
It is named by taking the path the executable and replacing the
'.exe' with '.log'

feerrenrut commented:
The system test failure happens due to the added logging. An error log message is sent to stdErr if logging handler has not been initialized. Py2exe watches stdErr and tries to write a log to the same directory as the NVDA executable, then shows a dialog. The error is a timeout waiting to acquire the mutex. Locally, the timeout situation is easy to reproduce, once installed, a second NVDA process will be started with the --ease-of-access argument whenever NVDA is started. This second process can not acquire the mutex. This behavior is not new. Due to the lack of logging it wasn't obvious. An overhaul of logging is required to be able to log these messages, otherwise the processes will overwrite the shared log file output name. This work for log files will come in a separate PR targeting 2021.2.

feerrenrut commented:
These new commits will be squashed. I'll leave them in place during secondary review so it is easier to see what has changed.

seanbudd commented on source/nvda.pyw:
```diff
@@ -213,12 +236,82 @@ def getDesktopName():
+				_log.debug(f"Failed waiting for mutex, error: {waitError}")
+				exception = winUser.WinError(waitError)
+			releaseResult = ctypes.windll.kernel32.ReleaseMutex(_mutex)
+			if 0 == releaseResult:
+				releaseError = winUser.GetLastError()
+				_log.debug(f"Failed to release mutex, error: {releaseError}")
+			closeResult = ctypes.windll.kernel32.CloseHandle(_mutex)
+			if 0 == closeResult:
+				closeError = winUser.GetLastError()
+				_log.debug(f"Failed to close mutex handle, error: {closeError}")
```
nitpicky - but should this repeated code to release and close mutex be abstracted into a function? I've replied to the duplicate section as well.

feerrenrut replied:
Potentially, note they are currently using different loggers. Given the time it takes to test this, I think I'll leave it as is.

seanbudd commented on source/nvda.pyw:
```diff
@@ -272,7 +365,20 @@ finally:
 		winUser.setSystemScreenReaderFlag(False)
-	ctypes.windll.kernel32.CloseHandle(mutex)
+
+	# From MS docs; "Multiple processes can have handles of the same mutex object"
+	# > Use the CloseHandle function to close the handle.
+	# > The system closes the handle automatically when the process terminates.
+	# > The mutex object is destroyed when its last handle has been closed.
+	# https://docs.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-createmutexw
+	releaseResult = ctypes.windll.kernel32.ReleaseMutex(mutex)
+	if 0 == releaseResult:
+		releaseError = winUser.GetLastError()
+		log.debug(f"Failed to release mutex, error: {releaseError}")
+	res = ctypes.windll.kernel32.CloseHandle(mutex)
+	if 0 == res:
+		error = winUser.GetLastError()
+		log.error(f"Unable to close mutex handle, last error: {winUser.WinError(error)}")
```
See other comment

seanbudd reviewed (approved):
Changes look good, I've left a couple of notes but this should be merged if they aren't an issue.

feerrenrut commented on source/gui/startupDialogs.py:
```diff
@@ -192,8 +192,10 @@ def onContinueRunning(self, evt):
+	def onExit(self, evt: wx.CommandEvent):
+		if not core.triggerNVDAExit():
+			log.error("NVDA already in process of exiting, this indicates a logic error.")
+		self.Destroy()  # Without this, the onExit is called multiple times by wx.
```
`triggerNVDAExit` only queues the `doShutdown` function. Many of the other callbacks call `Destroy` directly, so if doing that is a problem, there are plenty here. 
An alternative would be to `Unbind` the event handler within `onExit` so it can't be called a second time.

feerrenrut replied:
`onExit` was being called multiple times, adding `wx.Destroy` resolves this, likely because all event bindings are deleted.

#### First time cause
When the cancel button is pressed.

#### Second time
I suspect due to the cancel button using `wx.ID_CANCEL`. I think the `wx.Dialog` uses this as a default action when the dialog is closed (eg via close button).
Sequence:
- `triggerNVDAExit` queues `_doShutdown`.
- Later, when `_doShutdown` is run, it calls `_closeAllWindows`
- `_closeAllWindows` uses `wx.CallAfter` to close top level windows
- During `wx.Yield` the windows are closed, calling the `onExit` event handler for the `LauncherDialog`

feerrenrut committed:
Merge remote-tracking branch 'origin/master' into fixRestartAfterInstall

feerrenrut commented:
I haven't seen any reports that this issue still occurs on beta after #12538 was merged, we'll take this to beta via PR #12548.
